storage/xtradb/handler/handler0alter.cc

```diff
@@ -201,6 +201,49 @@ innobase_fulltext_exist(
 
 	return(false);
 }
+/*******************************************************************//**
+Determine if ALTER TABLE needs to rebuild the table .
+@param ha_alter_info		the DDL operation
+@param old_table			old table
+@return whether it is necessary to rebuild the table */
+static __attribute__((nonnull, warn_unused_result))
+bool
+innobase_need_rebuild_for_hash_cols(
+/*==================*/
+    Alter_inplace_info*	ha_alter_info,
+    dict_table_t*			old_table)
+{
+	/*If one of the hash indexes is dropped then
+	hash column associated with it needs to be dropped*/
+	for (uint i = 0; i < ha_alter_info->index_drop_count; i++) {
+		const KEY*	key
+			= ha_alter_info->index_drop_buffer[i];
+		dict_index_t*	index
+			= dict_table_get_index_on_name_and_min_id(
+				old_table, key->name);
+
+		if (index && index->type & DICT_UNIQUE_HASH) {
+			ha_alter_info->handler_flags |= Alter_inplace_info::DROP_COLUMN;
+			return (true);
+		}
+	}
+
+	/*If  hash index is added then
+	hash column associated with it will be added.*/
+	if ((ha_alter_info->handler_flags
+		    & Alter_inplace_info::ADD_UNIQUE_INDEX)) {
+		for (ulint i = 0; i < ha_alter_info->index_add_count; i++) {
+			const KEY* key = &ha_alter_info->key_info_buffer[
+				ha_alter_info->index_add_buffer[i]];
+			if(key->algorithm == HA_KEY_ALG_HASH){
+				ha_alter_info->handler_flags |= Alter_inplace_info::ADD_COLUMN;
+				return (true);
```

Comment:
Ok, this could be also correct. Please test alter table that (1) adds a blob column, adds a unique index to that column and (2) drop unique index, drop blob field that had a unique index.